odoo/addons/base/models/ir_cron.py

```diff
@@ -291,82 +306,147 @@ def _acquire_one_job(cls, cr, job_ids):
+            WHERE id = %s
+        """, [
+            job['failures'],
+            job['first_failure_date'],
+            job['active'],
+            job['id'],
+        ])
 
-        with cls.pool.cursor() as job_cr:
-            lastcall = fields.Datetime.to_datetime(job['lastcall'])
-            interval = _intervalTypes[job['interval_type']](job['interval_number'])
-            env = api.Environment(job_cr, job['user_id'], {'lastcall': lastcall})
-            ir_cron = env[cls._name]
-
-            # Use the user's timezone to compare and compute datetimes,
-            # otherwise unexpected results may appear. For instance, adding
-            # 1 month in UTC to July 1st at midnight in GMT+2 gives July 30
-            # instead of August 1st!
-            now = fields.Datetime.context_timestamp(ir_cron, datetime.utcnow())
-            past_nextcall = fields.Datetime.context_timestamp(
-                ir_cron, fields.Datetime.to_datetime(job['nextcall']))
-
-            # Compute how many call were missed
-            missed_call = past_nextcall
-            missed_call_count = 0
-            while missed_call <= now:
-                missed_call += interval
-                missed_call_count += 1
-            future_nextcall = missed_call
-
-            # Compute how many time we should run the cron
-            effective_call_count = (
-                     1 if not missed_call_count                    # run at least once
-                else 1 if not job['doall']                         # run once for all
-                else missed_call_count if job['numbercall'] == -1  # run them all
-                else min(missed_call_count, job['numbercall'])     # run maximum numbercall times
-            )
-            call_count_left = max(job['numbercall'] - effective_call_count, -1)
+        if completed:
+            ir_cron._mark_as_done(cron_cr, job)
+        elif os.getenv('ODOO_NOTIFY_CRON_CHANGES'): # See: `_notifydb`
+            ir_cron._cr.postcommit.add(ir_cron._notifydb)
 
-            # The actual cron execution
-            for call in range(effective_call_count):
-                ir_cron._callback(job['cron_name'], job['ir_actions_server_id'], job['id'])
+        cron_cr.commit()
 
-        # Update the cron with the information computed above
-        cron_cr.execute("""
+    @classmethod
+    def _execute_job(cls, job, timed_out_counter=0):
+        """ Execute the job's server action in a loop until it completes.
+        (i.e. the server action reports no remaining records to process)
+        or it fails without any record processed (i.e. the server action crashes and
+        reports no record done)
+        Each call to the ir_action has its own transaction and ir_cron_progress.
+        """
+        with cls.pool.cursor() as job_cr:
+            env = api.Environment(job_cr, job['user_id'], {'lastcall': job['lastcall'], 'cron_id': job['id']})
+            ir_cron = env[cls._name].browse(job['id'])
```

Comment:
Be careful of variable naming. `ir_cron` is acceptable when it only represents the model, but not when it represents a record. I would call it `cron` instead.

Reply:
Don't blame him, that's legacy code. There's even a `self.pool.check_signaling()` :see_no_evil:. I agree though that if we are moving the line then we can use the opportunity to rewrite a few variable here and there.